File: Sources/HerdrMobile/Transport/EventsSession.swift
import Foundation

/// Bounded backoff for events-channel reconnects (#18): capped exponential
/// delay, unlimited attempts. The Console keeps trying forever and shows
/// staleness through the session's status; the user decides when to give up.
struct ReconnectPolicy: Sendable, Equatable {
    var initialDelay: Duration
    var multiplier: Int
    var maxDelay: Duration

    static let `default` = ReconnectPolicy(
        initialDelay: .seconds(1), multiplier: 2, maxDelay: .seconds(30))

    /// The delay before reconnect attempt `attempt` (1-based): the initial
    /// delay grown by `multiplier` per prior failure, clamped at `maxDelay`.
    func delay(beforeAttempt attempt: Int) -> Duration {
        let factor = max(multiplier, 1)
        var delay = initialDelay
        var remaining = attempt - 1
        while remaining > 0, delay < maxDelay {
            delay = delay * factor
            remaining -= 1
        }
        return min(delay, maxDelay)
    }
}

/// Keepalive for the events session (#18). Citadel 0.12.1 exposes no
/// SSH-level keepalive (no ignore-packet or global-request API in it or its
/// NIOSSH fork) and no path to the NIO channel for TCP keepalive socket
/// options (`SSHClient.session` is internal; the `channelHandlers:` connect
/// parameter is stored but never installed). So the session pings herdr over
/// the ordinary RPC path instead — which is also the stronger check: it
/// generates SSH traffic that keeps NAT mappings alive, is bounded by the
/// per-request deadline, and exercises the whole path (SSH + socat + herdr),
/// so a dead connection is detected within interval + request timeout.
struct KeepalivePolicy: Sendable, Equatable {
    /// Idle time between pings while the events channel is live.
    var interval: Duration

    static let `default` = KeepalivePolicy(interval: .seconds(30))
}

/// Where the events session stands; the UI derives staleness from this.
enum EventsSessionStatus: Sendable, Equatable {
    /// The events channel is live. Emitted on every (re)connect — herdr does
    /// not replay state on subscribe (#4), so each `.connected` is the
    /// consumer's signal to re-snapshot via `listAgents()`.
    case connected
    /// The stream is down and attempt `attempt` starts after `delay`;
    /// everything shown since the last `.connected` may be stale. `failure`
    /// is what killed the previous attempt, for actionable user guidance.
    case reconnecting(attempt: Int, delay: Duration, failure: TransportError)
    /// Deliberately torn down (app backgrounded); no reconnect activity
    /// until `resume()`.
    case suspended
    /// `end()` was called; the session is finished for good.
    case ended
}

/// One element of the session's update stream: status transitions and events
/// interleaved in the order they happened.
enum EventsSessionUpdate: Sendable, Equatable {
    case status(EventsSessionStatus)
    case event(HerdrEvent)
}

/// The self-healing events channel for one Host (#18): owns the Host's
/// Transport, keeps its dedicated `events.subscribe` channel alive across
/// network blips and foreground/background transitions, and reports every
/// transition so the UI can show staleness.
///
/// Layer-honest reconnect: a dropped *channel* re-subscribes on the same SSH
/// connection through the transport's single-channel state machine; a dead
/// *SSH connection* (`isConnected == false`, or a timeout that means the
/// connection cannot be trusted) is closed and re-established via `connect`,
/// then pinged — the first call on every new connection path — before
/// re-subscribing.
///
/// Lifecycle: a fresh session is suspended; `resume()` activates it (call on
/// launch and on foregrounding), `suspend()` tears the channel and the SSH
/// connection down deliberately (call on backgrounding — iOS suspends
/// sockets anyway, an explicit close makes resume cheap and deterministic),
/// `end()` is terminal. All teardown is by explicit close, never by
/// cancelling a live exec channel (ADR 0002). Lifecycle calls serialize
/// internally — a `resume()` racing into a `suspend()`'s in-flight teardown
/// waits for it instead of interleaving (quick background→foreground
/// bounces are routine on iOS; callers never need to serialize their own
/// calls).
///
/// `updates` supports a single consumer and buffers without bound, exactly
/// like the underlying event stream (fine foregrounded in M0).
actor EventsSession {
    private enum Phase {
        case suspended, active, ended
    }

    /// Status transitions and events, in order. Finishes after `end()`.
    nonisolated let updates: AsyncStream<EventsSessionUpdate>
    private let updatesContinuation: AsyncStream<EventsSessionUpdate>.Continuation

    private let subscriptions: [EventSubscription]
    private let connect: @Sendable () async throws -> any Transport
    private let reconnectPolicy: ReconnectPolicy
    private let keepalive: KeepalivePolicy?

    private var phase: Phase = .suspended
    /// The Host's live Transport; consumers run snapshot RPCs (`listAgents`)
    /// through it after each `.connected`. Nil while suspended or between
    /// SSH re-establishments.
    private(set) var currentTransport: (any Transport)?
    /// Set when the connection can no longer be trusted even though it may
    /// still look alive (a timed-out request or keepalive ping): the next
    /// reconnect replaces the transport instead of reusing it.
    private var transportSuspect = false
    /// The keepalive failure that forced the current teardown, surfaced in
    /// the following `.reconnecting` status.
    private var pendingKeepaliveFailure: TransportError?
    private var liveStream: HerdrEventStream?
    private var runTask: Task<Void, Never>?
    private var keepaliveTask: Task<Void, Never>?
    private var backoffSleep: Task<Void, any Error>?
    /// The most recently enqueued lifecycle transition; each new one chains
    /// behind it, so transitions never interleave across the suspension
    /// points inside a teardown (see the actor doc).
    private var lifecycleTransition: Task<Void, Never>?

    init(
        subscriptions: [EventSubscription],
        connect: @escaping @Sendable () async throws -> any Transport,
        reconnectPolicy: ReconnectPolicy = .default,
        keepalive: KeepalivePolicy? = .default
    ) {
        self.subscriptions = subscriptions
        self.connect = connect
        self.reconnectPolicy = reconnectPolicy
        self.keepalive = keepalive
        (updates, updatesContinuation) = AsyncStream.makeStream(of: EventsSessionUpdate.self)
    }

    /// Activates the session (initially, or after `suspend()`): establishes
    /// the transport and channel, emitting `.connected` on success and
    /// `.reconnecting` transitions until then. Returns once any in-flight
    /// teardown has finished and the activation is underway. No-op while
    /// active or ended.
    func resume() async {
        await enqueueLifecycleTransition { await self.activate() }
    }

    /// Deliberate teardown for backgrounding: ends the events channel by
    /// explicit close, closes the SSH connection, and stops all reconnect
    /// activity. Returns once everything is down. No-op unless active.
    func suspend() async {
        await enqueueLifecycleTransition { await self.deactivate() }
    }

    /// Terminal teardown: like `suspend()`, then finishes `updates` for
    /// good. Idempotent.
    func end() async {
        await enqueueLifecycleTransition { await self.finish() }
    }

    private func activate() {
        guard phase == .suspended else { return }
        phase = .active
        runTask = Task { await self.run() }
    }

    private func deactivate() async {
        guard phase == .active else { return }
        phase = .suspended
        await windDown()
        updatesContinuation.yield(.status(.suspended))
    }

    private func finish() async {
        guard phase != .ended else { return }
        phase = .ended
        await windDown()
        updatesContinuation.yield(.status(.ended))
        updatesContinuation.finish()
    }

    /// Chains `transition` behind the previously enqueued one and waits for
    /// it. Enqueueing is synchronous on the actor, so the chain order is the
    /// call order and no transition ever observes another one mid-teardown.
    private func enqueueLifecycleTransition(
        _ transition: @escaping @Sendable () async -> Void
    ) async {
        let previous = lifecycleTransition
        let task = Task {
            await previous?.value
            await transition()
        }
        lifecycleTransition = task
        await task.value
    }

    // MARK: Reconnect loop

    /// One activation's lifetime: connect/subscribe, stream, and reconnect
    /// with bounded backoff, until the phase leaves `.active`. Exactly one
    /// run loop exists at a time — `suspend()`/`end()` await its exit before
    /// returning, and only `resume()` starts one.
    private func run() async {
        var attempt = 0
        while phase == .active {
            let stream: HerdrEventStream
            do {
                let transport = try await ensureTransport()
                guard phase == .active else { break }
                stream = try await transport.subscribeToEvents(subscriptions)
            } catch {
                guard phase == .active else { break }
                let failure = Self.transportFailure(error)
                if failure == .timedOut {
                    // The connection swallowed a request whole; do not trust
                    // it for the retry even if it still looks alive.
                    transportSuspect = true
                }
                attempt += 1
                await emitReconnectingAndBackOff(attempt: attempt, failure: failure)
                continue
            }
            if phase != .active {
                await stream.end()
                break
            }
            liveStream = stream
            attempt = 0
            pendingKeepaliveFailure = nil
            updatesContinuation.yield(.status(.connected))
            startKeepalive(stream: stream)

            var streamFailure: TransportError?
            do {
                for try await event in stream.events {
                    updatesContinuation.yield(.event(event))
                }
                // Graceful finish: the channel was ended explicitly, by
                // suspend()/end() or by a failed keepalive.
            } catch {
                streamFailure = Self.transportFailure(error)
            }
            stopKeepalive()
            liveStream = nil
            guard phase == .active else { break }
            let failure =
                streamFailure ?? pendingKeepaliveFailure
                ?? .channelFailed(detail: "events stream ended unexpectedly")
            pendingKeepaliveFailure = nil
            attempt += 1
            await emitReconnectingAndBackOff(attempt: attempt, failure: failure)
        }
    }

    /// The Host's live transport: reuses the current one while its SSH
    /// connection is alive and trusted, otherwise closes it and establishes
    /// a fresh one — pinged first, as on every new connection path.
    private func ensureTransport() async throws -> any Transport {
        if let transport = currentTransport {
            if !transportSuspect, await transport.isConnected {
                return transport
            }
            currentTransport = nil
            try? await transport.close()
        }
        let transport = try await connect()
        do {
            _ = try await transport.ping()
        } catch {
            try? await transport.close()
            throw error
        }
        transportSuspect = false
        currentTransport = transport
        return transport
    }

    private func emitReconnectingAndBackOff(attempt: Int, failure: TransportError) async {
        let delay = reconnectPolicy.delay(beforeAttempt: attempt)
        updatesContinuation.yield(
            .status(.reconnecting(attempt: attempt, delay: delay, failure: failure)))
        let sleep = Task { try await Task.sleep(for: delay) }
        backoffSleep = sleep
        try? await sleep.value
        backoffSleep = nil
    }

    /// Ends the current activation and closes everything: interrupts a
    /// backoff wait, ends the channel by explicit close, awaits the run
    /// loop's exit, then closes the SSH connection.
    private func windDown() async {
        backoffSleep?.cancel()
        stopKeepalive()
        if let stream = liveStream {
            await stream.end()
        }
        if let task = runTask {
            await task.value
            runTask = nil
        }
        if let transport = currentTransport {
            currentTransport = nil
            try? await transport.close()
        }
        transportSuspect = false
        pendingKeepaliveFailure = nil
    }

    // MARK: Keepalive

    private func startKeepalive(stream: HerdrEventStream) {
        guard let keepalive, let transport = currentTransport else { return }
        keepaliveTask = Task {
            while !Task.isCancelled {
                try? await Task.sleep(for: keepalive.interval)
                if Task.isCancelled { break }
                do {
                    _ = try await transport.ping()
                } catch is CancellationError {
                    break
                } catch TransportError.cancelled {
                    break
                } catch {
                    await self.keepaliveDidFail(Self.transportFailure(error), on: stream)
                    break
                }
            }
        }
    }

    private func stopKeepalive() {
        keepaliveTask?.cancel()
        keepaliveTask = nil
    }

    /// A keepalive ping failed: the connection cannot be trusted. Tears the
    /// stream down by explicit close; the run loop then reconnects with a
    /// fresh transport and surfaces this failure in `.reconnecting`.
    private func keepaliveDidFail(_ failure: TransportError, on stream: HerdrEventStream) async {
        guard phase == .active, liveStream === stream else { return }
        transportSuspect = true
        pendingKeepaliveFailure = failure
        await stream.end()
    }

    private static func transportFailure(_ error: any Error) -> TransportError {
        (error as? TransportError) ?? .channelFailed(detail: String(describing: error))
    }
}

File: android/core/src/main/java/dev/bybee/heeler/core/transport/EventsSession.kt
package dev.bybee.heeler.core.transport

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.delay
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlin.time.Duration
import kotlin.time.Duration.Companion.nanoseconds
import kotlin.time.Duration.Companion.seconds
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * Bounded backoff for transient events-channel failures: capped exponential
 * delay with unlimited attempts. Action-required failures stop and surface
 * their exact cause instead of retrying forever.
 */
data class ReconnectPolicy(
    val initialDelay: Duration,
    val multiplier: Int,
    val maxDelay: Duration,
) {
    companion object {
        val DEFAULT = ReconnectPolicy(1.seconds, 2, 30.seconds)
    }

    /** Delay before one-based [attempt]. */
    fun delayBeforeAttempt(attempt: Int): Duration {
        var delay = initialDelay
        repeat((attempt - 1).coerceAtLeast(0)) {
            if (delay < maxDelay) delay = (delay * multiplier.coerceAtLeast(1)).coerceAtMost(maxDelay)
        }
        return delay.coerceAtMost(maxDelay)
    }
}

/**
 * Events keepalive policy. The transport seam deliberately exposes no
 * SSH-specific heartbeat: ping over ordinary RPC exercises SSH, forwarding,
 * and herdr, keeps NAT mappings fresh, and is deadline-bounded.
 */
data class KeepalivePolicy(val interval: Duration) {
    companion object {
        val DEFAULT = KeepalivePolicy(30.seconds)
    }
}

/** Where the events session stands; UI derives staleness from this. */
sealed interface EventsSessionStatus {
    /** A live events channel; every connection requires a snapshot re-sync. */
    data object Connected : EventsSessionStatus
    /** The next retry starts after [delay]; state may be stale. */
    data class Reconnecting(
        val attempt: Int,
        val delay: Duration,
        val failure: TransportError,
    ) : EventsSessionStatus
    /** Reconnecting cannot repair this failure. */
    data class Failed(val failure: TransportError) : EventsSessionStatus
    /** Explicit background teardown. */
    data object Suspended : EventsSessionStatus
    /** [EventsSession.end] is terminal. */
    data object Ended : EventsSessionStatus
}

/** Status transitions and remote events interleaved in arrival order. */
sealed interface EventsSessionUpdate {
    data class Status(val status: EventsSessionStatus) : EventsSessionUpdate
    data class Event(val event: HerdrEvent) : EventsSessionUpdate
}

/**
 * The self-healing events channel for one Host. It owns the Host Transport,
 * keeps its dedicated `events.subscribe` channel across network blips and
 * foreground/background changes, and reports transitions so the UI can show
 * staleness.
 *
 * Layer-honest reconnect: a dropped channel re-subscribes on its live SSH
 * connection. A dead SSH connection (or a timed-out request that cannot be
 * trusted) is closed and established through [connect], then pinged first on
 * its new connection path before re-subscribing.
 *
 * Pane-scoped subscriptions are snapshot-derived and must never outlive their
 * connection: one exited pane makes `events.subscribe` all-or-nothing fail
 * with `pane_not_found`. [dropPaneSubscriptions] keeps one stale pane from
 * wedging a Host offline forever.
 */
class EventsSession(
    subscriptions: List<EventSubscription>,
    private val connect: suspend () -> Transport,
    private val reconnectPolicy: ReconnectPolicy = ReconnectPolicy.DEFAULT,
    private val keepalive: KeepalivePolicy? = KeepalivePolicy.DEFAULT,
    updatesBufferLimit: Int = HerdrEventStream.BUFFER_LIMIT,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Default),
) {
    private enum class Phase { SUSPENDED, ACTIVE, ENDED }

    /**
     * Ordered updates for one consumer. The buffer is bounded; overflow emits
     * [HerdrEvent.eventsDropped] so consumers re-snapshot rather than trusting
     * incomplete deltas.
     */
    val updates: Flow<EventsSessionUpdate>

    /** Latest measured ping latency; it is telemetry, not ordered convergence. */
    private val mutableLatencyUpdates = MutableSharedFlow<Duration>(replay = 0, extraBufferCapacity = 1)
    val latencyUpdates: Flow<Duration> = mutableLatencyUpdates

    private val updatesRelay = EventsUpdateRelay(updatesBufferLimit)
    private val stateMutex = Mutex()
    private val lifecycleMutex = Mutex()
    private val terminalMutex = Mutex()

    private var subscriptions: List<EventSubscription> = subscriptions
    private var phase = Phase.SUSPENDED
    private var currentTransport: Transport? = null
    /** Monotonic identity for installed transports; subscription reconnects reuse it. */
    var transportGeneration: ULong = 0u
        private set
    private var hasEstablishedTransport = false
    /** A successful events stream makes a later transient socket-open failure recoverable. */
    private var hasEstablishedEventsStream = false
    private var transportSuspect = false
    private var pendingKeepaliveFailure: TransportError? = null
    private var lastConnectionActivityNanos: Long? = null
    private var resubscribeRequested = false
    private var liveStream: HerdrEventStream? = null
    private var activationGeneration: ULong = 0u
    private var runJob: Job? = null
    private var keepaliveJob: Job? = null

    init {
        require(updatesBufferLimit > 0) { "Events updates buffer must be positive." }
        updates = updatesRelay.flow
    }

    /** Activates a fresh or suspended session. No-op while active or ended. */
    suspend fun resume() {
        lifecycleMutex.withLock { activateIfSuspended() }
    }

    /** Restarts a session after an actionable failure or explicit user retry. */
    suspend fun retry() {
        lifecycleMutex.withLock {
            if (stateMutex.withLock { phase == Phase.ACTIVE }) {
                deactivate(emitStatus = false)
            }
            activateIfSuspended()
        }
    }

    /**
     * Re-proves an apparently live session when foregrounding. Frozen sockets
     * do not necessarily end their event flow; a failed ping feeds the normal
     * reconnect sequence instead of waiting for the keepalive interval.
     */
    suspend fun revalidate() {
        val (stream, transport) = stateMutex.withLock {
            if (phase != Phase.ACTIVE) return
            liveStream to currentTransport
        }
        if (stream == null || transport == null) return
        try {
            mutableLatencyUpdates.tryEmit(measureLatency(transport))
            noteConnectionActivity()
        } catch (failure: CancellationException) {
            throw failure
        } catch (failure: Throwable) {
            keepaliveDidFail(asTransportFailure(failure), stream)
        }
    }

    /** Deliberate background teardown: ends channel and SSH connection. */
    suspend fun suspendSession() = lifecycleMutex.withLock { deactivate(emitStatus = true) }

    /** Terminal teardown. The session cannot be resumed after this. */
    suspend fun end() = lifecycleMutex.withLock {
        val shouldFinish = stateMutex.withLock {
            if (phase == Phase.ENDED) false else {
                phase = Phase.ENDED
                true
            }
        }
        if (!shouldFinish) return@withLock
        windDown()
        yieldUpdate(EventsSessionUpdate.Status(EventsSessionStatus.Ended))
        updatesRelay.finish()
        scope.cancel()
    }

    /**
     * Replaces subscriptions. A live channel ends explicitly and re-subscribes
     * immediately on the same transport; its ensuing Connected update requires
     * the ordinary snapshot re-sync. Pane entries still disappear on every
     * disconnect and must be reinstalled after each Connected update.
     */
    suspend fun updateSubscriptions(newSubscriptions: List<EventSubscription>) {
        val stream = stateMutex.withLock {
            if (subscriptions == newSubscriptions) return
            subscriptions = newSubscriptions
            if (phase == Phase.ACTIVE && liveStream != null) {
                resubscribeRequested = true
                liveStream
            } else {
                null
            }
        }
        stream?.end()
    }

    suspend fun <T> withTransport(operation: suspend (Transport) -> T): T {
        val transport = stateMutex.withLock {
            currentTransport ?: throw TransportError.SshUnreachable("The Host is not connected.")
        }
        val value = operation(transport)
        noteConnectionActivity()
        return value
    }

    /**
     * Runs one terminal lifetime with exclusive access to the Host terminal
     * channel. The permit spans explicit terminal teardown.
     */
    suspend fun <T> withTerminalTransport(operation: suspend (Transport) -> T): T =
        terminalMutex.withLock { withTransport(operation) }

    /** A test/diagnostic count of updates shed by the bounded buffer. */
    val droppedUpdateCount: Long get() = updatesRelay.droppedCount

    private suspend fun deactivate(emitStatus: Boolean) {
        val shouldDeactivate = stateMutex.withLock {
            if (phase != Phase.ACTIVE) false else {
                phase = Phase.SUSPENDED
                true
            }
        }
        if (!shouldDeactivate) return
        windDown()
        if (emitStatus) yieldUpdate(EventsSessionUpdate.Status(EventsSessionStatus.Suspended))
    }
    private suspend fun activateIfSuspended() {
        val generation = stateMutex.withLock {
            if (phase != Phase.SUSPENDED) return
            phase = Phase.ACTIVE
            activationGeneration++
            activationGeneration
        }
        stateMutex.withLock {
            runJob = scope.launch { run(generation) }
        }
    }

    private suspend fun run(generation: ULong) {
        var attempt = 0
        while (activationIsCurrent(generation)) {
            val stream = try {
                val transport = ensureTransport(generation)
                val requested = stateMutex.withLock { subscriptions }
                transport.subscribeToEvents(requested)
            } catch (failure: Throwable) {
                val transportFailure = asTransportFailure(failure)
                if (
                    transportFailure == TransportError.TimedOut ||
                    transportFailure is TransportError.StreamLocalOpenFailed
                ) {
                    stateMutex.withLock { transportSuspect = true }
                }
                if (namesMissingPane(transportFailure) && dropPaneSubscriptions()) continue
                if (!retryableForSession(transportFailure)) {
                    yieldUpdate(EventsSessionUpdate.Status(EventsSessionStatus.Failed(transportFailure)))
                    return
                }
                dropPaneSubscriptions()
                attempt++
                reconnectAfter(attempt, transportFailure, generation)
                continue
            }

            if (!activationIsCurrent(generation)) {
                stream.end()
                return
            }
            stateMutex.withLock {
                liveStream = stream
                hasEstablishedEventsStream = true
                pendingKeepaliveFailure = null
            }
            attempt = 0
            yieldUpdate(EventsSessionUpdate.Status(EventsSessionStatus.Connected))
            startKeepalive(stream)

            var streamFailure: TransportError? = null
            try {
                stream.events.collect { event ->
                    if (activationIsCurrent(generation)) {
                        noteConnectionActivity()
                        yieldUpdate(EventsSessionUpdate.Event(event))
                    }
                }
            } catch (failure: Throwable) {
                if (failure !is CancellationException) streamFailure = asTransportFailure(failure)
            }

            val deliberateResubscribe: Boolean
            val keepaliveFailure: TransportError?
            stateMutex.withLock {
                if (liveStream === stream) {
                    stopKeepaliveLocked()
                    liveStream = null
                }
                deliberateResubscribe = resubscribeRequested
                resubscribeRequested = false
                keepaliveFailure = pendingKeepaliveFailure
                pendingKeepaliveFailure = null
            }
            if (!activationIsCurrent(generation)) return
            if (deliberateResubscribe) continue

            val failure = keepaliveFailure ?: streamFailure
                ?: TransportError.ChannelFailed("events stream ended unexpectedly")
            if (!retryableForSession(failure)) {
                yieldUpdate(EventsSessionUpdate.Status(EventsSessionStatus.Failed(failure)))
                return
            }
            dropPaneSubscriptions()
            attempt++
            reconnectAfter(attempt, failure, generation)
        }
    }

    /**
     * Discards pane-scoped entries while retaining globals. herdr rejects a
     * whole `events.subscribe` if one pane has exited; carrying snapshot-derived
     * entries over a reconnect would retry that doomed set forever.
     */
    private suspend fun dropPaneSubscriptions(): Boolean = stateMutex.withLock {
        val globals = subscriptions.filterIsInstance<EventSubscription.Global>()
        if (globals.size == subscriptions.size) false else {
            subscriptions = globals
            true
        }
    }

    private fun namesMissingPane(failure: TransportError): Boolean =
        failure is TransportError.ApiRejected && failure.code == "pane_not_found"

    private suspend fun ensureTransport(generation: ULong): Transport {
        if (!activationIsCurrent(generation)) throw TransportError.Cancelled
        val current = stateMutex.withLock { currentTransport to transportSuspect }
        if (current.first != null && !current.second && current.first!!.isConnected()) {
            return current.first!!
        }
        current.first?.let {
            stateMutex.withLock { if (currentTransport === it) currentTransport = null }
            try {
                it.close()
            } catch (_: Throwable) {
                // A replacement connection cannot use an old transport either way.
            }
        }

        val fresh = connect()
        try {
            if (!activationIsCurrent(generation)) throw TransportError.Cancelled
            mutableLatencyUpdates.tryEmit(measureLatency(fresh))
            if (!activationIsCurrent(generation)) throw TransportError.Cancelled
        } catch (failure: Throwable) {
            try {
                fresh.close()
            } catch (_: Throwable) {
                // The connection is already rejected; preserve the original failure.
            }
            throw failure
        }
        stateMutex.withLock {
            transportSuspect = false
            currentTransport = fresh
            if (hasEstablishedTransport) transportGeneration++ else hasEstablishedTransport = true
        }
        noteConnectionActivity()
        return fresh
    }

    private suspend fun reconnectAfter(attempt: Int, failure: TransportError, generation: ULong) {
        val backoffDelay = reconnectPolicy.delayBeforeAttempt(attempt)
        yieldUpdate(
            EventsSessionUpdate.Status(
                EventsSessionStatus.Reconnecting(attempt, backoffDelay, failure),
            ),
        )
        delay(backoffDelay)
        currentCoroutineContext().ensureActive()
        if (!activationIsCurrent(generation)) return
    }

    private data class WindDownResources(
        val run: Job?,
        val keepalive: Job?,
        val stream: HerdrEventStream?,
        val transport: Transport?,
    )

    private suspend fun windDown() {
        val resources = stateMutex.withLock {
            WindDownResources(runJob, keepaliveJob, liveStream, currentTransport).also {
                runJob = null
                keepaliveJob = null
                liveStream = null
                currentTransport = null
                transportSuspect = false
                pendingKeepaliveFailure = null
                resubscribeRequested = false
                lastConnectionActivityNanos = null
            }
        }
        resources.run?.cancel()
        resources.keepalive?.cancel()
        try {
            resources.transport?.close()
        } catch (_: Throwable) {
            // Explicit teardown is terminal even when the remote close races a disconnect.
        }
        resources.stream?.end()
        // Wait for any attach operation before a replacement transport can install.
        terminalMutex.withLock { }
        dropPaneSubscriptions()
    }

    private suspend fun startKeepalive(stream: HerdrEventStream) {
        val policy = keepalive ?: return
        val job = scope.launch {
            while (true) {
                delay(policy.interval)
                currentCoroutineContext().ensureActive()
                if (!connectionIsIdle(policy.interval)) continue
                val transport = stateMutex.withLock { currentTransport } ?: return@launch
                try {
                    mutableLatencyUpdates.tryEmit(measureLatency(transport))
                    noteConnectionActivity()
                } catch (failure: CancellationException) {
                    return@launch
                } catch (failure: Throwable) {
                    keepaliveDidFail(asTransportFailure(failure), stream)
                    return@launch
                }
            }
        }
        stateMutex.withLock {
            stopKeepaliveLocked()
            keepaliveJob = job
        }
    }

    private fun stopKeepaliveLocked() {
        keepaliveJob?.cancel()
        keepaliveJob = null
    }

    private suspend fun keepaliveDidFail(failure: TransportError, stream: HerdrEventStream) {
        val shouldEnd = stateMutex.withLock {
            if (phase != Phase.ACTIVE || liveStream !== stream) false else {
                transportSuspect = true
                pendingKeepaliveFailure = failure
                true
            }
        }
        if (shouldEnd) stream.end()
    }

    private suspend fun activationIsCurrent(generation: ULong): Boolean = stateMutex.withLock {
        phase == Phase.ACTIVE && activationGeneration == generation
    }

    private suspend fun measureLatency(transport: Transport): Duration {
        val started = System.nanoTime()
        transport.ping()
        return (System.nanoTime() - started).nanoseconds
    }

    private suspend fun noteConnectionActivity() {
        stateMutex.withLock { lastConnectionActivityNanos = System.nanoTime() }
    }

    private suspend fun connectionIsIdle(interval: Duration): Boolean = stateMutex.withLock {
        val last = lastConnectionActivityNanos ?: return@withLock true
        System.nanoTime() - last >= interval.inWholeNanoseconds
    }

    /**
     * A first stream-local failure is still terminal: it can mean disabled
     * forwarding or a wrong Host setup. Once this Host has already delivered
     * events, the same failure after a dropped SSH connection is indistinct
     * from herdr briefly restarting, so use normal bounded reconnect backoff.
     */
    private suspend fun retryableForSession(failure: TransportError): Boolean =
        failure.isRetryable || (
            failure is TransportError.StreamLocalOpenFailed &&
                stateMutex.withLock { hasEstablishedEventsStream }
            )

    private fun yieldUpdate(update: EventsSessionUpdate) = updatesRelay.yield(update)

    private fun asTransportFailure(failure: Throwable): TransportError = when (failure) {
        is TransportError -> failure
        is HerdrApiError -> TransportError.ApiRejected(failure.code, failure.serverMessage)
        is CancellationException -> TransportError.Cancelled
        else -> TransportError.ChannelFailed(failure.toString())
    }
}

/**
 * One-consumer bounded update relay. It owns overflow accounting because stock
 * DROP_OLDEST flows drop silently; every shed item becomes a local drop marker.
 */
private class EventsUpdateRelay(private val limit: Int) {
    private val lock = ReentrantLock()
    private val queue = ArrayDeque<EventsSessionUpdate>()
    private var waiter: kotlinx.coroutines.CompletableDeferred<Unit>? = null
    private var finished = false
    private var consumer: Job? = null
    var droppedCount: Long = 0
        private set

    val flow: Flow<EventsSessionUpdate> = flow {
        val reader = currentCoroutineContext()[Job] ?: error("Events flow requires a Job")
        claim(reader)
        while (true) {
            val next = next(reader) ?: break
            emit(next)
        }
    }

    fun yield(update: EventsSessionUpdate) {
        val wake = lock.withLock {
            if (finished) return
            if (update is EventsSessionUpdate.Status &&
                (update.status == EventsSessionStatus.Suspended || update.status == EventsSessionStatus.Ended)
            ) {
                queue.clear()
                queue.addLast(update)
            } else if (queue.size < limit) {
                queue.addLast(update)
            } else {
                queue.removeAt(0)
                droppedCount++
                queue.addLast(update)
                if (queue.size == limit) {
                    queue.removeAt(0)
                    droppedCount++
                }
                queue.addLast(EventsSessionUpdate.Event(HerdrEvent.eventsDropped))
            }
            waiter.also { waiter = null }
        }
        wake?.complete(Unit)
    }

    fun finish() {
        val wake = lock.withLock {
            finished = true
            waiter.also { waiter = null }
        }
        wake?.complete(Unit)
    }

    private fun claim(reader: Job) = lock.withLock {
        check(consumer == null || consumer === reader) { "Events updates support one consumer." }
        consumer = reader
    }

    private suspend fun next(reader: Job): EventsSessionUpdate? {
        while (true) {
            val waiting = lock.withLock {
                check(consumer === reader) { "Events updates are owned by another consumer." }
                queue.removeFirstOrNull()?.let { return it }
                if (finished) return null
                check(waiter == null) { "Events updates have more than one reader." }
                kotlinx.coroutines.CompletableDeferred<Unit>().also { waiter = it }
            }
            try {
                waiting.await()
            } catch (failure: CancellationException) {
                lock.withLock { if (waiter === waiting) waiter = null }
                throw failure
            }
        }
    }
}

